File: src/hyphae/view/viewer.py
"""What a route module is built with: the store to read, and the way a page is rendered.

`Viewer` is one per app. A route module takes it as its factory's argument rather than
reaching into `request.app.state`, so a route body stays a plain typed function of what it
needs.
"""

from dataclasses import dataclass
from pathlib import Path

from fastapi import Request
from fastapi.responses import Response
from fastapi.templating import Jinja2Templates


@dataclass(frozen=True)
class Viewer:
    """The store every route reads, and the environment it renders through."""

    db: Path
    templates: Jinja2Templates

    def error(self, request: Request, status: int, message: str) -> Response:
        """The error page, which is what every handler in `build_app` answers with."""
        return self.templates.TemplateResponse(
            request, "error.html", {"status": status, "message": message}, status_code=status
        )

File: rust/crates/hyphae-view/src/viewer.rs
//! What a route module is built with: the store to read, and whether `--dev` is on.
//!
//! Ported from `src/hyphae/view/viewer.py`. One per app, handed to each route as its state rather
//! than reached for through a global.

use std::path::PathBuf;

use crate::store::Reader;

/// The store every route reads, and whether `--dev` is on.
pub struct Viewer {
    pub reader: Reader,
    /// Under `--dev`, the directory the static files are read from — which is also the directory
    /// the reload loop watches ([`crate::dev`]). `None` is the shipped viewer, which answers from
    /// the copy compiled into the binary.
    pub dev: Option<PathBuf>,
}

impl Viewer {
    /// Whether this is the dev viewer, which is all a page needs to know: the one line it adds.
    #[must_use]
    pub fn dev(&self) -> bool {
        self.dev.is_some()
    }
}
